### ISSUE TITLE
manhole breaks on gtk.keysyms AttributeError

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2464]
|<img alt="gyst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| gyst reported|
|-|-|
|Trac ID|trac#2464|
|Type|defect|
|Created|2007-02-20 18:24:15Z|
|Branch|https://github.com/twisted/twisted/tree/manhole-2464|

The bug:
--------

```

gyst@sirius:~/downloads/Twisted-2.5.0$ manhole
Traceback (most recent call last):
  File "/usr/bin/manhole", line 21, in ?
    manhole.run()
  File "/usr/lib/python2.4/site-packages/twisted/scripts/manhole.py", line 102, in run
    run(config)
  File "/usr/lib/python2.4/site-packages/twisted/scripts/manhole.py", line 154, in run_gtk2
    w = gtk2manhole.ManholeWindow()
  File "/usr/lib/python2.4/site-packages/twisted/manhole/ui/gtk2manhole.py", line 46, in __init__
    self.input = ConsoleInput(self._input)
  File "/usr/lib/python2.4/site-packages/twisted/manhole/ui/gtk2manhole.py", line 218, in __init__
    self.rkeymap[getattr(gtk.keysyms, name)] = name
AttributeError: 'module' object has no attribute '__getattr__'
```

The fix:
--------

```

gyst@sirius:/usr/lib/python2.4/site-packages/twisted/manhole/ui$ diff -u gtk2manhole.orig gtk2manhole.py 
--- gtk2manhole.orig    2007-02-20 19:18:27.000000000 +0100
+++ gtk2manhole.py      2007-02-20 19:18:46.000000000 +0100
@@ -216,7 +216,7 @@
         for name in dir(gtk.keysyms):
             try:
                 self.rkeymap[getattr(gtk.keysyms, name)] = name
-            except TypeError:
+            except (TypeError, AttributeError):
                 pass
 
     def _on_key_press_event(self, entry, event):


```

<details><summary>Searchable metadata</summary>

```
trac-id__2464 2464
type__defect defect
reporter__gyst gyst
priority__normal normal
milestone__ 
branch__branches_manhole_2464 branches/manhole-2464
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__manhole__gtk2 manhole, gtk2
time__1171995855000000 1171995855000000
changetime__1237776497000000 1237776497000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="gyst's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|gyst commented|
|-|-|

Sorry, forgot to mention: this is on Ubuntu Linux with the following python gtk binding:

```
ii  python-gtk2       2.10.3-0ubuntu3   Python bindings for the GTK+ widget set
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

This seems to be a new problem in the version of the Python GTK+ bindings in edgy.  Dapper packages 2.8.6-0ubuntu1, and the problem does not occur there.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

[#3337](https://github.com/twisted/twisted/issues/3337) has been closed as a duplicate of this one.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [26464]) Branching to 'manhole-2464'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph removed owner|
|-|-|

A fix, along with what I believe is the first GTK UI test in Twisted; certainly the first test for manhole.

You can find build results here:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/manhole-2464

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @glyph|
|-|-|

Test run moderately unsuccessful.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph removed owner|
|-|-|

Successful "reactors" run still pending, but I'm pretty confident it will work since I've tried it on a similarly configured Dapper box now (with every permutation of gtk1/gtk2/DISPLAY/no-DISPLAY).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra set owner to @glyph|
|-|-|

1.[twisted/manhole/ui/test/__init__.py](twisted/manhole/ui/test/__init__.py) should have a copyright header
 1. [twisted/manhole/ui/test/test_gtk2manhole.py](twisted/manhole/ui/test/test_gtk2manhole.py) should have a copyright header
 1. the updated copyright header in [twisted/manhole/ui/gtk2manhole.py](twisted/manhole/ui/gtk2manhole.py) should simply look like:
```
# Copyright (c) 2001-2009 Twisted Matrix Laboratories.
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph removed owner|
|-|-|

Copyright statements updated.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_10"></a>|@mwhudson set owner to @glyph|
|-|-|

Looks fine, please land.  The test is a bit assert x == x, but well, hard to do better.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set status to `closed`|
|-|-|

(In [26517]) Update the GTK2 "manhole" to work on versions of PyGTK which use lazy modules.

Author: glyph

Reviewer: exarkun, mwh

Fixes [#2464](https://github.com/twisted/twisted/issues/2464)

On most recent installations of pygtk, this makes the 'manhole' command-line
tool actually start the GUI rather than printing a traceback and immediately
exiting.

I this also adds what I believe is the first test coverage for any of the GUI
applications within Twisted.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
